### DISCUSSION TITLE
Pinging / Accessing WSL2 IP from a docker container

### DISCUSSION THREAD

**Discussion by addshore** [microsoft/WSL#7066] [Q&A] [answered]
I'm running WSL2 and docker-for-win.
In the past I have been able to have docker containers communicate with processes running in WSL, but it seems in the past week or so things have stopped working for me.

I have scoured the resources that I could find, lots of which refer to WSL -> Windows or docker containers -> Windows etc, but not often container -> WSL itself.
I have tried poking various firewalls (normally testing by just turning them off etc), and tried poking a bunch of other stuff but really am starting to draw a blank.

My personal reproduction right now would be:
 - `ifconfig` ina  WSL shell to get the IP address of WSL, eg `172.19.65.61`
 - Create some docker container and try to ping the IP from that container
(I have also tried telneting to a specific port that had some service running)

Any reading or pointers that I may have overlooked, or misread etc would be greatly appreciated.

Perhaps this question would be better aimed toward docker-for-win instead of wsl, but I'v got to start somewhere!

-----

To make this issue even more searchable on the web I'll add that All of this was happening while trying to use xdebug for php in a docker container, connecting back to VSCode in remote mode on WSL2.
I was not able to create this debug connection, which I when I realized something was going on with networking.

People that are having this issue may just be trying to connect to the wrong IP / domain, and reading https://addshore.com/2021/02/vscode-php-xdebug-from-localhost-wsl2-and-docker-containers/ might help there.
Otherwise if having a similar subnet issue as me then nuking docker and reinstalling might help.

**Comment by addshore** [comment, marked as answer]
After trying a variety of debugging etc I ended up **uninstalling and reinstalling docker**, and everything seems to start working just fine again.

The issue was caused by subnet differences between WSL and docker.

By default, as far as I can tell WSL & Docker so something special together so that containers end up with an IP on the same subnet as WSL.
For some reason my setup was not doing this.
After reinstalling my containers end up with a bridge network again with 172. IPs (same as WSL). Prior to this the containers were getting 192. IPs.

The added confusion comes from the default docker subnet, and also seemingly configured docker subnet in the UI
![image](https://user-images.githubusercontent.com/3308769/123542751-3d877400-d743-11eb-9b28-3f3802e71bd1.png)

----

To make this issue even more searchable on the web I'll add that All of this was happening while trying to use xdebug for php in a docker container, connecting back to VSCode in remote mode on WSL2.
I was not able to create this debug connection, which I when I realized something was going on with networking.

People that are having this issue may just be trying to connect to the wrong IP / domain, and reading https://addshore.com/2021/02/vscode-php-xdebug-from-localhost-wsl2-and-docker-containers/ might help there.
Otherwise if having a similar subnet issue as me then nuking docker and reinstalling might help.

**Reply by saschanockel** [reply]
I am running almost the exact same setup, just with PhpStorm instead of VSCode. Finally after trying everything I could think of the thing that fixed it was reinstalling Docker.

I also want to note that this just stopped working from one day to the other and delayed my work by several hours. With all the problems I encountered over the time using Docker on Windows I really can't recommend it if you have the possibility to work with a native Linux instead.

**Reply by worthwhileindustries** [reply]
> After reinstalling my containers end up with a bridge network again with 172. IPs (same as WSL). Prior to this the containers were getting 192. IPs.

This wasn't specifically happening but, uninstalling, reboot, reinstalling fixed this issue after an insane amount of wasted time trying to figure out what the problem was.

**Reply by tomsit-ionos** [reply]
re-install work for me too! 
I believe i also had a subnet problem although it must have been the mask that was off, because both WSL as well as the container got an address with a leading 172 but still could not reach each other.

**Comment by RichardDavies** [comment]
Ugh. I just spent the better part of three days pulling my hair out trying to figure out why Xdebug wouldn't work anymore in my WSL2/Lando/Docker/PHP/VSCode environment. After much frustration I eventually I traced it down to an inability of my Docker container to access the WSL2 host IP where VS Code is running. I could ping other IPs from within Docker, but I couldn't ping my WSL2 IP...

I read a bunch of posts online about others with similar issues and tried everything from changing VSCode/Xdebug settings, disabling firewalls, and various internet/network troubleshooting steps, but nothing seemed to fix it or suggest any reason why my Docker container couldn't access my WSL2 host IP.

Finally, I found this post and in desperation uninstalled Docker and reinstalled it as suggested and voilà---Xdebug magically works again! It's frustrating that we still don't know why this sometimes happens to WSL2/Docker, but at least I know a relatively quick fix now for when it does happen again next time.

**Reply by worthwhileindustries** [reply]
@RichardDavies the only thing I can think of without actually having tested it is I updated WSL2 via `wsl --update` or via updates to windows and that may have messed up the networking for docker desktop integration.
